splicemachine/spliceengine
# DB-9876 Use HBase Memstore compaction to run MVCC vacuuming

## Short Description

MVCC vacuuming now runs during memstore compaction when `hbase.hregion.compacting.memstore.type` is set to `"EAGER"`

## Long Description

This makes use of [HBase in-memory compaction (Accordion)](https://blogs.apache.org/hbase/entry/accordion-hbase-breathes-with-in).
Up until now, the only way for us to run MVCC vacuuming was to wait for the memstore to be full. Upon flushing it to HFiles, we would remove deleted rows and old updates. When purging a lot of records, this could lead to flushing only a fraction of the cells that filled the memstore. Accordion allows us to run MVCC vacuuming within the memstore, so workloads that focus on updating the same rows would in theory never trigger a memstore flush.

## How to test

On a standalone instance of splice machine, with a targeted use case:
```
drop table foo;
create table foo (a int, b int);
insert into foo values (1,2);
update foo set b = 3;
update foo set b = 4;
update foo set b = 3;
update foo set b = 4;
...
-- 200000 more updates
...
update foo set b = 3;
update foo set b = 4;
update foo set b = 3;
update foo set b = 4;
```
We can notice that the memstore never flushes since memstore compaction maintains its size to a small number.
